src/main/java/net/neoforged/neoforge/registries/NeoForgeRegistryCallbacks.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.registries;
+
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import net.minecraft.core.IdMapper;
+import net.minecraft.core.Registry;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.world.entity.ai.attributes.Attribute;
+import net.minecraft.world.entity.ai.attributes.DefaultAttributes;
+import net.minecraft.world.entity.ai.village.poi.PoiType;
+import net.minecraft.world.item.BlockItem;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.levelgen.DebugLevelSource;
+import net.neoforged.neoforge.registries.callback.AddCallback;
+import net.neoforged.neoforge.registries.callback.BakeCallback;
+import net.neoforged.neoforge.registries.callback.ClearCallback;
+
+class NeoForgeRegistryCallbacks {
+    static class BlockCallbacks implements ClearCallback<Block>, BakeCallback<Block> {
+        static final BlockCallbacks INSTANCE = new BlockCallbacks();
+        static final ClearableObjectIntIdentityMap<BlockState> BLOCKSTATE_TO_ID_MAP = new ClearableObjectIntIdentityMap<>();
+
+        @Override
+        public void onClear(Registry<Block> registry, boolean full) {
+            BLOCKSTATE_TO_ID_MAP.clear();
+        }
+
+        @Override
+        public void onBake(Registry<Block> registry) {
+            for (Block block : registry) {
+                for (BlockState state : block.getStateDefinition().getPossibleStates()) {
+                    BLOCKSTATE_TO_ID_MAP.add(state);
+                    state.initCache();
```

Comment:
We need to research whether `initCache` really needs to be called again when baking the registry; it is a huge performance bottleneck when launching the game/joining servers. I am 90% sure this hasn't been needed for a while.

Reply:
Seems useless indeed. Removed that and also `getLootTable`.